modules/core/shared/src/main/scala/net/protocol/BindExecute.scala

```diff
@@ -0,0 +1,186 @@
+            } *>
+            new CopyNotSupportedException(statement.command).raiseError[F, Completion]
+
+          case CopyInResponse(_) =>
+            send(CopyFail) *>
+            expect { case ErrorResponse(_) => } *>
+            send(Sync) *>
+            expect { case ReadyForQuery(_) => } *>
+            new CopyNotSupportedException(statement.command).raiseError[F, Completion]
+
+          case ErrorResponse(info) =>
+            for {
+              hi <- history(Int.MaxValue)
+              _  <- send(Sync)
+              _  <- expect { case ReadyForQuery(_) => }
+              redactedArgs = statement.command.encoder.types.zip(
+                redactionStrategy.redactArguments(statement.command.encoder.encode(args)))
+              a  <- new PostgresErrorException(
+                      sql             = statement.command.sql,
+                      sqlOrigin       = Some(statement.command.origin),
+                      info            = info,
+                      history         = hi,
+                      arguments       = redactedArgs,
+                      argumentsOrigin = Some(argsOrigin)
+                    ).raiseError[F, Completion]
+            } yield a
+        }
+
+        Resource.make {
+          exchange("bind+execute"){ (span: Span[F]) =>
+            for {
+              pn <- preBind(span)
+              _  <- send(ExecuteMessage(pn.value, 0))
+              _  <- send(Flush)
+              _  <- postBind
+              c  <- postExec
+            } yield new Protocol.CommandPortal[F, A](pn, statement, args, argsOrigin) {
+              def execute: F[Completion] = c.pure
+            }
+          }
+        } { portal => Close[F].apply(portal.id)}
+
+      }
+
+      def query[A, B](
+        statement:  Protocol.PreparedQuery[F, A, B],
+        args:       A,
+        argsOrigin: Origin,
+        redactionStrategy: RedactionStrategy,
+        initialSize: Int
+      ): Resource[F, Protocol.QueryPortal[F, A, B]] = {
+        val (preBind, postBind) = bindExchange(statement, args, argsOrigin, redactionStrategy)
+        Resource.eval(Deferred[F, Unit]).flatMap { prefetch =>
+          Resource.make {
+            exchange("bind+execute"){ (span: Span[F]) =>
+              for {
+                pn <- preBind(span)
+                _  <- span.addAttributes(
+                        Attribute("max-rows",  initialSize.toLong),
+                        Attribute("portal-id", pn.value)
+                      )
+                _  <- send(ExecuteMessage(pn.value, initialSize))
+                _  <- send(Flush)
+                _  <- postBind
+                rs <- unroll(statement, args, argsOrigin, redactionStrategy)
+              } yield new Protocol.QueryPortal[F, A, B](pn, statement, args, argsOrigin, redactionStrategy) {
+                def execute(maxRows: Int): F[List[B] ~ Boolean] = 
+                  prefetch.tryGet.flatMap {
+                    case None => rs.pure <* prefetch.complete(())
```

Comment:
Is it safe to ignore `maxRows` here?

Reply:
Two use case of this :
* `PreparedQuery.fetch2` : bind with size 2, execute with size 2
* `PreparedQuery.stream`: bind with size `batchSize`, execute with the same size

I resorted to ignore the parameter instead of separating QueryPortal into several cases: one for fixed size query without continuation (fetch2), one for fixed size with continuation (stream), and one for arbitrary sized execute (cursor).

This can be done in a follow up PR to cleanup the internals a bit :)

Reply:
Moreover, I separated the case `bindSized` from `bind` at the protocol prepared query level. This is clearer to see if we are in the BingExecute case or classic Bind then Execute.